pkg/metrics/metrics.go

```diff
@@ -0,0 +1,254 @@
+package metrics
+
+import (
+	"context"
+	"time"
+
+	log "github.com/Azure/aad-pod-identity/pkg/logger"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// This const block defines the metric names.
+const (
+	assignedIdentityAdditionDurationName                = "assigned_identity_addition_duration_seconds"
+	assignedIdentityAdditionCountName                   = "assigned_identity_addition_count"
+	assignedIdentityDeletionDurationName                = "assigned_identity_deletion_duration_seconds"
+	assignedIdentityDeletionCountName                   = "assigned_identity_deletion_count"
+	nodeManagedIdentityOperationsDurationName           = "nodemanagedidentity_operations_duration_seconds"
+	managedIdentityControllerCycleDurationName          = "managedidentitycontroller_cycle_duration_seconds"
+	managedIdentityControllerCycleCountName             = "managedidentitycontroller_cycle_count"
+	managedIdentityControllerNewLeaderElectionCountName = "managedidentitycontroller_new_leader_election_count"
+	cloudProviderOperationsErrorsCountName              = "cloud_provider_operations_errors_count"
+	cloudProviderOperationsDurationName                 = "cloud_provider_operations_duration_seconds"
+	kubernetesAPIOperationsErrorsCountName              = "kubernetes_api_operations_errors_count"
+)
+
+// The following variables are measures
+var (
+
+	// AssignedIdentityAdditionDurationM is a measure that tracks the duration in seconds of assigned_identity_addition operations.
+	AssignedIdentityAdditionDurationM = stats.Float64(
+		assignedIdentityAdditionDurationName,
+		"Duration in seconds of assigned identity addition operations",
+		stats.UnitMilliseconds)
+
+	// AssignedIdentityAdditionCountM is a measure that tracks the cumulative number of assigned identity addition operations.
+	AssignedIdentityAdditionCountM = stats.Int64(
+		assignedIdentityAdditionCountName,
+		"Total number of assigned identity addition operations",
+		stats.UnitDimensionless)
+
+	// AssignedIdentityDeletionDurationM is a measure that tracks the duration in seconds of assigned_identity_deletion operations.
+	AssignedIdentityDeletionDurationM = stats.Float64(
+		assignedIdentityDeletionDurationName,
+		"Duration in seconds of assigned identity deletion operations",
+		stats.UnitMilliseconds)
+
+	// AssignedIdentityDeletionCountM is a measure that tracks the cumulative number of assigned identity deletion operations.
+	AssignedIdentityDeletionCountM = stats.Int64(assignedIdentityDeletionCountName,
+		"Total number of assigned identity deletion operations",
+		stats.UnitDimensionless)
+
+	// NodeManagedIdentityOperationsDurationM is a measure that tracks the duration in seconds of nodemanagedidentity operations.
+	NodeManagedIdentityOperationsDurationM = stats.Float64(
+		nodeManagedIdentityOperationsDurationName,
+		"Duration in seconds of node managed identity operations",
+		stats.UnitMilliseconds)
+	// "operation_type", "status_code"
```

Comment:
Can we add some labels to this metric? I would suggest the following:
- AzureIdentity name (k8s object) which related to this request
- K8s namespace of the AzureIdentity 
- Azure resource ID (of which we are trying to get token for)

We are dealing with several namespaces in every cluster, and expect a lot of different identities defined. These labels can help us a lot with managing all of these objects. 
WDYT?

Reply:
@danielShor - Agree with your inputs.

Discussed metric NodeManagedIdentityOperationsDurationM, is being recorded in function ServeHTTP (for both success as well as error paths) in server.go(nmi). In this context we don't have details of identity but we do have reference of *http.Request and following details can be retrieved.
- NameSpace : Can be retrieved from header : "podns"
- AzureResource: Can be retrieved from query parameters : key :"resource"

Let me know if we can proceed only with these two new labels ?

Reply:
@surenderssm, these 2 are great. Can we also get the AzureIdentity name (k8s object) which related to this request? (I know it might be complicated, but can be very helpful). If too complicated for now, and may take some time we will be happy with these 2 labels :)

Reply:
@danielshor - Added namespace,resource label to NodeManagedIdentityOperationsDurationM metric. We will analyse and decide on AzureIdentity label in future iterations.